### ISSUE TITLE
Support addAllProperty

### ISSUE THREAD

**Issue by wenerme** [mapstruct/mapstruct#1343]
Mapstruct 1.2 will detect `addProperty` method, if adder is prefered, will use this instead of `getList().add()`, this is very useful, e.g. jpa entity add children, or some builder only got adder method.

So, I suggest support `addAllProperty` instead of `getList().addAll()` or for loop `addProperty`, sometime, the builder can only `addAllProperty`, because `getList()` return an immutable view.

If there is a `addAllProperty`, it means "please use this method instead".

**Comment by filiphr** [comment]
Thanks for creating this @wenerme. I've added it to the future planning. If someone is interested contributing this we can certainly help out ')

**Comment by filiphr** [comment]
There is also an SO [question](https://stackoverflow.com/questions/48658716/mapstruct-protobuf3-mapping-a-map/48673581#48673581) for a similar feature

**Comment by anthonycorbacho** [comment]
I found a workaround by doing something like 

```java
@Mapping(target = "mutableProfiles", source = "profiles"),
//Since mapping is done in mutableProfiles we need to tell mapstruct to not try to map this.
@Mapping(target = "profiles", ignore = true),
```

The first line will use the mutable map and add the collection and the 2nd will tell mapstruct to not try to map profile since its already done.

**Comment by alex-lzl** [comment]
I can't find any property with "mutable" prefix in generated Protobuf java classes. Currently have to skip the mapping and use a @AfterMapping class (1.3.0beta1). Hopefully this feature will be implemented soon! Thank you!

**Comment by filiphr** [comment]
Can some of you perhaps provide a small example where this is created by protobuf? That way it might be easier for us to have a look at this.

**Comment by vordimous** [comment]
I have the example repo. I can add a map mapping example to it. Ill try to do it this weekend or first thing Monday.

**Comment by vordimous** [comment]
I pushed an updated example to my [Protobuf-Mapstruct repo](https://github.com/vordimous/mapstruct-proto)

Here is what is being generated by mapstruct:
``` java
if ( userDTO.getMetadata() != null ) {
	Map<String, String> map = user.getMetadata();
	if ( map != null ) {
		userDTO.getMetadata().putAll( map );
	}
}
```

Here is what should be generated:
``` java
if ( userDTO.getMetadata() != null ) {
	Map<String, String> map = user.getMetadata();
	if ( map != null ) {
		userDTO.putAllMetadata( map );
	}
}
```

This will meet the new Protobuf Api and not throw errors.

**Comment by vordimous** [comment]
@filiphr is that what you needed?

**Comment by vsr625** [comment]
Any updates on this issue?

**Comment by nanda-XITE** [comment]
The issue for mapping from a List<String> to a repeated string field is still present. I would love to submit a PR if you could point me in the right direction @filiphr

**Comment by filiphr** [comment]
Thanks for the help @nanda-XITE, I think that you can start by looking into [`Type#getAdderForType`](https://github.com/mapstruct/mapstruct/blob/119826982ab7a8d7d5a995ca80319992acca150f/processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java#L673).

That place is where the potential adder is used. `Type#getAdders` returns all the adders. The `DefaultAccessorNamingStrategy` would return that `addAllX` is an adder. However, I think that this might need some changes for `Map`. Also look into `AdderWrapper` and `AdderWrapper.ftl`, this would also need changing in case an adder with a collection is used.

**Comment by azimoff98** [comment]
i am facing with this issue right now and unfortunately cannot take advantages of your comments :( please help

**Comment by gigi** [comment]
Hi guys! Any updates?

**Comment by PiotrKupniewski** [comment]
Hey, am facing the same issue problem, did someone fix that ? ;)

**Comment by PiotrKupniewski** [comment]
This solution works for me : 
```

Mapping(target = "mutableProfiles", source = "profiles"),
//Since mapping is done in mutableProfiles we need to tell mapstruct to not try to map this.
@Mapping(target = "profiles", ignore = true),
```

**Comment by sakhno** [comment]
Hi, just faced the same issue with protobaf 3.19.3 and mapstruct 1.4.2:

> java.lang.UnsupportedOperationException: null
> 	at java.base/java.util.Collections$UnmodifiableMap.putAll(Collections.java:1463)

Are there any workarounds?

**Comment by tomas--** [comment]
This workaround worked for me. For collection field use method with `@BeforeMapping` annotation and don't forget to ignore that field using `@Mapping(target = "mapField", ignore = true)` to preserve mapstruct generate wrong code for protobuf use case.
```
  @BeforeMapping
  default void wireMap(Source sourceObject, @MappingTarget Target.Builder builder) {
    builder.putAllMapField(sourceObject.getMapField());
  }

```

**Comment by sakhno** [comment]
We ended up on defining our own map wrappers like:

```
message MapOfStrings {
    map<string, string> values = 1;
}
```

**Comment by tomas--** [comment]
Yes that's also option, but it makes your proto model more complex.

**Comment by ninjachen** [comment]
Any update?
I still meet this issue on `org.mapstruct:mapstruct -> 1.4.1.Final` + protobuf
It generates `builder.getxxxxx().addAll()`, and it should use `builder.addAllxxxxx()`

The workaround is copy the codes generated, and modify it as I wish

**Comment by AndyChenIT** [comment]
The same issue,  [version: 1.5.0.beta1], `builder.getxxxxx().addAll()` will throw UnsupportedOperationException.

**Comment by aviaur** [comment]
Hi guys! Any updates?

**Comment by filiphr** [comment]
@ninjachen and @AndyChenIT you can always use `@Mapper(collectionMappingStrategy = CollectionMappingStrategy.TARGET_IMMUTABLE)` and then the `getXXXX()` will not be used, neither do the `addAllxxx` will be used as well.

Since there is a MapStruct extension for Protobuf supported by out community in https://github.com/entur/mapstruct-spi-protobuf I would suggest trying to add such a support out-of-the-box there.
